File: ensarkbank-android/app/src/main/java/com/ensark/ensarkbank/repository/TransactionRepository.java
package com.ensark.ensarkbank.repository;

import android.content.Context;

import com.ensark.ensarkbank.api.ApiClient;
import com.ensark.ensarkbank.api.TransactionApiService;
import com.ensark.ensarkbank.model.dto.AccountTransactionRequest;
import com.ensark.ensarkbank.model.dto.AccountTransactionResponse;
import com.ensark.ensarkbank.model.dto.ATMTransactionRequest;
import com.ensark.ensarkbank.model.dto.ATMTransactionResponse;
import com.ensark.ensarkbank.model.dto.BalanceCheckRequest;
import com.ensark.ensarkbank.model.dto.JournalResponse;
import com.ensark.ensarkbank.model.dto.OtpInitiateResponse;
import com.ensark.ensarkbank.model.dto.OtpVerifyRequest;

import java.math.BigDecimal;
import java.util.List;

import okhttp3.ResponseBody;
import retrofit2.Callback;

public class TransactionRepository {

    private final TransactionApiService apiService;

    public TransactionRepository(Context context) {
        apiService = ApiClient.getClient(context).create(TransactionApiService.class);
    }

    public void initiateOnlineTransaction(AccountTransactionRequest atr, Callback<OtpInitiateResponse> callback) {
        apiService.initiateOnlineTransaction(atr).enqueue(callback);
    }

    public void verifyOnlineTransaction(OtpVerifyRequest req, Callback<AccountTransactionResponse> callback) {
        apiService.verifyOnlineTransaction(req).enqueue(callback);
    }

    public void getTransactionById(Long id, Callback<AccountTransactionResponse> callback) {
        apiService.getTransactionById(id).enqueue(callback);
    }

    public void findTransactionsByAccountNumber(String accountNumber, Callback<List<AccountTransactionResponse>> callback) {
        apiService.findTransactionsByAccountNumber(accountNumber).enqueue(callback);
    }

    public void findTransactionsByAccountId(Long accountId, Callback<List<AccountTransactionResponse>> callback) {
        apiService.findTransactionsByAccountId(accountId).enqueue(callback);
    }

    public void findJournalsByCustomerEmail(String email, Callback<List<JournalResponse>> callback) {
        apiService.findJournalsByCustomerEmail(email).enqueue(callback);
    }

    public void findJournalsByEmailAndSpan(String email, String from, String to, Callback<List<JournalResponse>> callback) {
        apiService.findJournalsByEmailAndSpan(email, from, to).enqueue(callback);
    }

    public void exportStatement(String accountNumber, String from, String to, String format, Callback<ResponseBody> callback) {
        apiService.exportStatement(accountNumber, from, to, format).enqueue(callback);
    }

    public void atmTransaction(ATMTransactionRequest request, Callback<ATMTransactionResponse> callback) {
        apiService.atmTransaction(request).enqueue(callback);
    }

    public void checkAtmBalance(BalanceCheckRequest request, Callback<BigDecimal> callback) {
        apiService.checkAtmBalance(request).enqueue(callback);
    }

    public void getAtmTransactionsByCardNumber(String cardNumber, Callback<List<ATMTransactionResponse>> callback) {
        apiService.getAtmTransactionsByCardNumber(cardNumber).enqueue(callback);
    }
}

File: ensark-flutter/lib/repositories/transaction_repository.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';
import '../models/transaction/transaction_models.dart';
import '../models/card/card_models.dart';
import '../providers/core_providers.dart';

part 'transaction_repository.g.dart';

class TransactionRepository {
  final Dio _dio;

  TransactionRepository(this._dio);

  Future<OtpInitiateResponse> initiateOnlineTransaction(AccountTransactionRequest request) async {
    final response = await _dio.post('api/account-transaction/online/initiate', data: request.toJson());
    return OtpInitiateResponse.fromJson(response.data);
  }

  Future<AccountTransactionResponse> verifyOnlineTransaction(OtpVerifyRequest request) async {
    final response = await _dio.post('api/account-transaction/online/verify', data: request.toJson());
    return AccountTransactionResponse.fromJson(response.data);
  }

  Future<AccountTransactionResponse> getTransactionById(int id) async {
    final response = await _dio.get('api/account-transaction/$id');
    return AccountTransactionResponse.fromJson(response.data);
  }

  Future<List<AccountTransactionResponse>> findTransactionsByAccountNumber(String accountNumber) async {
    final response = await _dio.get('api/account-transaction/accountNumber/$accountNumber');
    return (response.data as List).map((e) => AccountTransactionResponse.fromJson(e)).toList();
  }

  Future<List<JournalResponse>> findJournalsByCustomerEmail(String email) async {
    final response = await _dio.get('api/journal/customer/$email');
    return (response.data as List).map((e) => JournalResponse.fromJson(e)).toList();
  }

  Future<List<JournalResponse>> findJournalsByEmailAndSpan(String email, String from, String to) async {
    final response = await _dio.get('api/journal/customer/$email/span', queryParameters: {'from': from, 'to': to});
    return (response.data as List).map((e) => JournalResponse.fromJson(e)).toList();
  }

  Future<List<int>> exportStatement(String accountNumber, String from, String to, String format) async {
    final response = await _dio.get(
      'api/journal/account/$accountNumber/export',
      queryParameters: {'from': from, 'to': to, 'format': format},
      options: Options(responseType: ResponseType.bytes),
    );
    return response.data as List<int>;
  }

  Future<double> checkAtmBalance(BalanceCheckRequest request) async {
    final response = await _dio.post('api/atm-transactions/balance', data: request.toJson());
    return (response.data as num).toDouble();
  }

  Future<ATMTransactionResponse> atmTransaction(ATMTransactionRequest request) async {
    final response = await _dio.post('api/atm-transactions', data: request.toJson());
    return ATMTransactionResponse.fromJson(response.data);
  }
}

@riverpod
TransactionRepository transactionRepository(Ref ref) {
  return TransactionRepository(ref.watch(dioProvider));
}
